fairytale.cpp

```diff
@@ -151,57 +152,63 @@ int main(int argc, char** argv) {
 #ifdef WINDOWS
   _setmaxstdio(2048);
 #endif
-  printf("Fairytale Prototype v0.016 by M. Pais, 2018\n\n");
-  if (argc < 4 || argc > 7 || strlen(argv[1]) != 4 || (!isdigit(argv[1][0])) || (!isdigit(argv[1][1])) || (argv[1][0] > argv[1][1]) || ((argv[1][2] & 0xDF) != 'B') || ((argv[1][3] & 0xDF) != 'D')) {
-    printf("Syntax: fairytale MTBD output_file input_files\n");
-    printf("M: Memory cache coefficient, [0..9] -> [4MB..2048MB]\n");
-    printf("T: Total storage coefficient, [0..9] -> [8MB..4096MB]\n");
-    printf("B: Brute force DEFLATE streams [B -> on, b-> off]\n");
-    printf("D: Perform deduplication stage [D -> on, d-> off]\n");
-    printf("M<=T\n");
-    getchar();
-    return 0;
-  }
+  CLI::App app{"Fairytale Prototype v0.016 by M. Pais, 2018"};
+  int memory = 9;
+  int total_storage = 9;
+  bool brute_mode = false;
+  bool deduplication = false;
+  bool verbose = false;
+  std::string output_file;
+  std::vector<std::string> input_files;
+  app.add_option("-m,--memory", memory, "Memory cache coefficient, [0..9] -> [4MB..2048MB]", 9);
+  app.add_option("-t,--total-storage", total_storage, "Total storage coefficient, [0..9] -> [8MB..4096MB]", 9);
+  app.add_flag("-b,--brute", brute_mode, "Brute force DEFLATE streams");
+  app.add_flag("-d,--deduplication", deduplication, "Perform deduplication stage");
+  app.add_flag("-v,--verbose", verbose, "Enable verbose output");
+  app.add_option("output_file,-o", output_file, "output_file")->required();
+  app.add_option("input_files,-i,", input_files, "input_files")->required();
+  CLI11_PARSE(app, argc, argv);
+
   clock_t start_time = clock();
   FileStream output;
-  output.create(argv[2]);
+  output.create(output_file.c_str());
 
-  Array<FileStream*> input(argc - 3);
+  Array<FileStream*> input(input_files.size());
   Block* block = new Block{ 0 };
   Block* root = block;
-  for (int i = 3; i < argc; i++) {
-    input[i - 3] = new FileStream;
-    if (!input[i - 3]->open(argv[i])) {
-      printf("File not found: %s\n", argv[i]);
+  for (size_t i = 0; i < input_files.size(); i++) {
+    input[i] = new FileStream;
+    if (!input[i]->open(input_files[i].c_str())) {
+      printf("File not found: %s\n", input_files[i].c_str());
       getchar();
       return 0;
     }
-    block->data = input[i - 3];
+    block->data = input[i];
     block->length = block->data->getSize();
-    printf("Loaded %s, (%" PRIu64 " bytes), hashing... ", argv[i], block->length);
+    printf("Loaded %s, (%" PRIu64 " bytes), hashing... ", input_files[i].c_str(), block->length);
     block->calculateHash();
     printf("done\n");
-    if (i > 3)
-      input[i - 3]->goToSleep();
-    if (i + 1 < argc) {
+    if (i > 0)
+      input[i]->goToSleep();
+    if (i + 1 < input_files.size()) {

```

Comment:
I think this logic might be wrong. Should it be `if (i - 1 < input_files.size()) {`?

Reply:
Don't think so. As far as I can tell, this checks if we're not at the end of the `for` loop - the condition will be false with the last index i (which is `input_files.size() - 1`). `i + 1` will be `input_files.size()` in this case, so both will be equal, the condition gets false and there will be no next block created.

An alternative would be `if (i != input_files.size() - 1)`. Also, you could add a comment like `// don't do this for the last iteration of i`

The logic is the same to the code before because you changed the start value for `i`, but not its end value which just was renamed (`input_files.size()` instead of `argc`)